### PR TITLE
Adding module and state to download artifacts from Nexus Repository

### PR DESCRIPTION
**Pull request by anitakrueger**
### What does this PR do?
This PR adds a module to connect to Nexus Repository like the artifactory module does. It has been copied from the artifactory module and adapted to fit Nexus 3.0.
The accompanying state provides one state to make sure artifacts from Nexus are downloaded to a given location.
Unfortunately the artifactory module can't be used to talk to Nexus because their repository structure is different.

### What issues does this PR fix or reference?
None

### New Behavior
Artifacts can be downloaded from Nexus based on being the latest snapshot or a specific release version.

### Tests written?
No

### CODE REVIEW

**Comment by anitakrueger** [comment]
I'll look at the failed checks tonight.

**Comment by cachedout** [comment]
Hi @anitakrueger 

Glad to see you! We are moving toward a policy where all new modules/states need to have accompanying test coverage to be merged into Salt. Otherwise, they'll have to go into the `contrib` repo. Could you add tests for these? Thanks.

**Comment by anitakrueger** [comment]
Hey Mike, yes of course. Makes sense. I need to work on those test failures anyways. Might have to wait until the weekend when I get to it.

**Comment by rallytime** [comment]
re-run py2

**Comment by rallytime** [comment]
Hi @anitakrueger - good to see you! I am re-running the Ubuntu tests. Those failures didn't look related to your change to me, so we'll see how this run goes. That being said, I think the only one you need to worry about right now is the `codeclimate` result.

File: salt/modules/nexus.py [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+            result['status'] = False
+            result['comment'] = checksum_comment
+            return result
+
+    log.debug('Downloading: {url} -> {target_file}'.format(url=artifact_url, target_file=target_file))
+
+    try:
+        request = urllib.request.Request(artifact_url, None, headers)
+        f = urllib.request.urlopen(request)
+        with salt.utils.fopen(target_file, "wb") as local_file:
```
Comment (rallytime) [inline salt/modules/nexus.py]: This has been moved to `salt.utils.files.fopen`. Can you update this reference?

File: salt/states/nexus.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+'''
+This state downloads artifacts from Nexus 3.x.
+
+'''
```
Comment (rallytime) [inline salt/states/nexus.py]: A version added tag would be good here as well.

File: salt/states/nexus.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+# Import python libs
+from __future__ import absolute_import
+import logging
+
+log = logging.getLogger(__name__)
+
+
+def downloaded(name, artifact, target_dir='/tmp', target_file=None):
```
Comment (rallytime) [inline salt/states/nexus.py]: This state should have a `__virtual__()` function to gate loading it as well.

**Review by rallytime — CHANGES_REQUESTED** [review]
Overall, this looks great. I have a couple of minor comments that should be addressed.

File: salt/modules/nexus.py [hunk 4]
```diff
@@ -0,0 +1,501 @@
+# -*- coding: utf-8 -*-
+'''
+Module for fetching artifacts from Nexus 3.x
+'''
```
[the remark this answers is not in the document]
Comment (anitakrueger) [reply salt/modules/nexus.py]: Fixed

File: salt/modules/nexus.py [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+
+# Import python libs
+from __future__ import absolute_import
+import os
+import base64
+import logging
+
+# Import Salt libs
+import salt.utils
```
[the remark this answers is not in the document]
Comment (anitakrueger) [reply salt/modules/nexus.py]: Fixed

File: salt/modules/nexus.py (hunk 1 above)
Comment (anitakrueger) [reply salt/modules/nexus.py]: Fixed

File: salt/states/nexus.py (hunk 2 above)
Comment (anitakrueger) [reply salt/states/nexus.py]: Fixed

File: salt/states/nexus.py (hunk 3 above)
Comment (anitakrueger) [reply salt/states/nexus.py]: Added __virtual__ function

**Comment by anitakrueger** [comment]
I'll need a couple more days to add those tests for the module and state. Bit of a learning curve for me, but I'll look at what is currently there for artifactory as this is what nexus was based on.

**Comment by cachedout** [comment]
@anitakrueger Sounds good. Keep us posted. Thanks!

**Comment by anitakrueger** [comment]
Added the tests and ran them using
```
python tests/runtests.py -n unit.states.test_nexus
```
and
```
python tests/runtests.py -n unit.modules.test_nexus
```
All tests were successful.

**Comment by rallytime** [comment]
@anitakrueger Awesome! Those new tests are running on our Jenkins runs too. 👌 

The last thing to clean up here before a final review are a couple of pylint warnings: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/16315/violations/file/salt/states/nexus.py/

Do you have a moment to clean those up? After that, we can update the branch and that should get the Py3 tests happy, too. (They were broken downstream for a bit, but we've fixed them now.)

**Comment by anitakrueger** [comment]
@rallytime Sorry, it looks like I made a mistake when rebasing develop onto my new branch :( I did fix the pylint errors. Argh. How do I fix this please?

**Comment by anitakrueger** [comment]
Looks like I need a lesson in github branching :( I did a pull upstream on my develop branch and then a rebase onto the add-feature-nexus branch. I am sure I did this before and it didn't look like this.

**Comment by rallytime** [comment]
Hmm...I am not sure offhand what happened here, but it looks like maybe you rebased develop against your fork instead of saltstack's. I usually do something like this (assuming the saltstack repo is `upstream`, and not your github fork):
```
git fetch upstream
git checkout develop
git reset --hard upstream/develop
git checkout add-feature-nexus
git rebase develop
```
Does that help?

**Comment by anitakrueger** [comment]
@rallytime something must have gone wrong with my branch :(
even the clean rebase produces a conflict now.
I'm opening a new clean PR. Sorry for all this mess.

**Comment by anitakrueger** [comment]
Closed in favor of https://github.com/saltstack/salt/pull/44591

**Comment by rallytime** [comment]
@anitakrueger No problem at all! Thanks for the addition. :)

**Pull request closed without merging**